Sources/ClientRuntime/Networking/Http/AsyncThrowingStreamReader.swift

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import AwsCommonRuntimeKit
+
+/// A `StreamReader` that reads `AsyncThrowingStream<Data, Error>`
+/// Event Stream messages are sent as a series of `Data` chunks
+public class AsyncThrowingStreamReader: StreamReader {
+    public var availableForRead: UInt {
+        get {
+            fatalError("This operation is not supported")
+        }
+        set {
+            fatalError("This operation is not supported")
+        }
+    }
+
+    private var _hasFinishedWriting: Bool
+    public var hasFinishedWriting: Bool {
+        get {
+            lock.withLockingClosure {
+                return _hasFinishedWriting
+            }
+        }
+        set {
+            lock.withLockingClosure {
+                _hasFinishedWriting = newValue
+            }
+        }
+    }
+
+    internal let stream: AsyncThrowingStream<Data, Error>
+    internal var buffer: Data = Data()
+    private let lock = NSLock()
+
+    public init(stream: AsyncThrowingStream<Data, Error>) {
+        self.stream = stream
+        self._hasFinishedWriting = false
+    }
+
+    public func read(maxBytes: UInt?, rewind: Bool = false) -> AwsCommonRuntimeKit.ByteBuffer {
+        if rewind {
+            fatalError("Rewind is not supported with async streams")
+        }
+
+        return lock.withLockingClosure {
+            return _unsafeWait {
+                return await self.read(maxBytes: maxBytes)
```

Comment:
See warning in the [NSLock docs](https://developer.apple.com/documentation/foundation/nslock):

> The NSLock class uses POSIX threads to implement its locking behavior. When sending an unlock message to an NSLock object, you must be sure that message is sent from the same thread that sent the initial lock message. Unlocking a lock from a different thread can result in undefined behavior.

It seems likely that after `await` this block will continue execution on a different thread.

Reply:
no longer applicable.

Reply:
How is this no longer applicable?  We still make an `async` call while holding a lock, and there is no guarantee that execution continues on the same thread after resuming.

Reply:
I made the comment after converting StreamReader to actor which didn't work in the end.

Reply:
Also, why it needs to be guaranteed that execution must continue on the same thread?

Reply:
While yes, acquiring the lock and releasing the lock must happen on the same thread, I don't see any threading issues here. 

`_unsafeWait` will execute the async closure in a task on some random thread, but `_unsafeAwait` will return the value in the same thread that it was called on. This is because `_unsafeWait` uses a semaphore to block the calling thread until the async task finishes.